File: c/employee_list_removal.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void display_employees(char *[], int);
void remove_employee(char *[], char *, int);
bool find_employee(char *[], char *, int);

char *names[] = {"John Smith", "Jackie Jackson", "Chris Jones", "Amanda Cullen",
                 "Jeremy Goodwin"};

int main(void) {
  int array_length = sizeof(names) / sizeof(names[0]);
  char remove[15];

  display_employees(names, array_length);

  printf("Enter an employee name to remove: ");
  fgets(remove, 15, stdin);
  remove[strcspn(remove, "\n")] = 0;

  /*if (find_employee(names, remove, initial_names_length)) {*/
  /*} else {*/
  /*  puts("No employee name found");*/
  /*}*/
  remove_employee(names, remove, array_length);

  printf("Exist: %d\n", find_employee(names, remove, array_length));

  array_length = sizeof(names) / sizeof(names[0]);

  display_employees(names, array_length);
  return EXIT_SUCCESS;
}

void display_employees(char *employees[], int len) {
  printf("There are %d employees:\n", len);
  for (int i = 0; i < len; i++) {
    printf("- %s\n", employees[i]);
  }
}

void remove_employee(char *array[], char *name, int length) {
  for (int i = 0; i < length; i++) {
    if (!strcmp(array[i], name)) {
      printf("found %s\n", name);
      array[i] = array[i + 1];
    }
  }
}

bool find_employee(char *employees[], char *name, int length) {
  for (int i = 0; i < length; i++) {
    if (!strcmp(employees[i], name)) {
      return true;
    }
  }

  return false;
}

File: zig/employee_list_removal.zig
const std = @import("std");

pub fn main() !void {
    const allocator = std.heap.page_allocator;

    const stdin = std.io.getStdIn().reader();
    var buffer: [1024]u8 = undefined;

    var employees_list = std.ArrayList([]const u8).init(allocator);
    defer employees_list.deinit();
    try employees_list.appendSlice(&[_][]const u8{
        "John Smith",
        "Jackie Jackson",
        "Chris Jones",
        "Amanda Cullen",
        "Jeremy Goodwin",
    });

    display_employees(&employees_list);

    std.debug.print("Enter an employee name to remove: ", .{});
    const name = try stdin.readUntilDelimiter(buffer[0..], '\n');

    remove_employee(&employees_list, name);

    display_employees(&employees_list);
}

fn remove_employee(list: *std.ArrayList([]const u8), name: []const u8) void {
    for (list.items, 0..) |value, i| {
        if (std.mem.eql(u8, value, name)) {
            _ = list.orderedRemove(i);
            break;
        }
    }

    std.debug.print("The name \"{s}\" does not exists", .{name});
}

fn display_employees(list: *std.ArrayList([]const u8)) void {
    std.debug.print("There are {d} employees:\n", .{list.items.len});
    for (list.items) |employee| {
        std.debug.print("- {s}\n", .{employee});
    }
}
